File: wordBreak2.java
//refs to http://www.cnblogs.com/feiling/p/3357067.html "dp + dfs"
public class Solution {
     public ArrayList<String> wordBreak(String s, Set<String> dict) {
        int n = s.length();
        ArrayList<String> ret = new ArrayList<String>();
        if (n == 0) return ret;
        boolean [][]ref = new boolean[n][n + 1];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= n; j++)
                ref[i][j] = false;
        }
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = 1; j <= n - i; j++) {
                if (dict.contains(s.substring(i, i + j))) {
                    ref[i][j] = true;
                    continue;
                }
                
                for (int k = 1; k < j; k++) {
                    if (ref[i][k] && ref[k + i][j - k]) ref[i][j] = true;
                }
            }
        }
        if (!ref[0][n]) return ret;
        StringBuffer str = new StringBuffer();
        Break(s, 0, str, dict, ret, ref);
        return ret;
    }
    
    public void Break(String s, int start, StringBuffer str, Set<String> dict, ArrayList<String> ret, boolean [][] ref) {
        if (start == s.length()) {
            String tmp = str.toString();
            ret.add(tmp.substring(0, tmp.length() - 1));
            return;
        }
        
        for (int len = 1; len <= s.length() - start; len++) {
            if (ref[start][len]) {
                if (dict.contains(s.substring(start, start + len))) {
                    int beforeAddlen = str.length();
                    str.append(s.substring(start, start + len) + " ");
                    Break(s, start + len, str, dict, ret, ref);
                    str.delete(beforeAddlen, str.length());
                }
            }
        }
        return;
    }
    //--------------------------------------------------------------//
     public ArrayList<String> wordBreak(String s, Set<String> dict) {
        ArrayList<String> ret = new ArrayList();
        int n = s.length();
        if (n == 0) return ret;
        boolean[][] ref = new boolean[n][n + 1];
        for (int len = 1; len <= n; len++) {
            for (int i = 0; i < n - len + 1; i++) {
                if (dict.contains(s.substring(i, i + len))) {
                    ref[i][len] = true;
                    continue;
                }
                
                for (int k = 1; k < len; k++) {
                    if (ref[i][k] && ref[i + k][len - k]) 
                        ref[i][len] = true;
                }
            }
        }
        if (ref[0][n]) dfs(s, dict, 0, 0, ret, new StringBuffer(), ref);
        
        return ret;
    }
    
    private void dfs (String s, Set<String> dict, int start, int depth, ArrayList<String> ret, StringBuffer sb, boolean[][] ref) 
    {
        if (depth == s.length()) {
            String tmp = sb.toString();
            ret.add(tmp.substring(0, tmp.length() - 1));
            return;
        }
        
        for (int i = 1; i <= s.length() - start; i++) {
            if (ref[start][i]) {
                if ( dict.contains(s.substring(start, start + i)) ) {
                    int beforeAddlen = sb.length();
                    sb.append(s.substring(start, start + i) + " ");
                    dfs(s, dict, start + i, start + i, ret, sb, ref);
                    sb.delete(beforeAddlen, sb.length());
                }
            }
        }
    }

    //--------------------------------------------------------------//
    public ArrayList<String> wordBreak(String s, Set<String> dict) {
        ArrayList<String> ret = new ArrayList<String>();
        if (s.length() == 0 ) return ret;
        int length = s.length();
        boolean[][] ref = new boolean[length][length + 1];
        
        for (int len = 1; len <= length; len++) {
            for (int i = 0; i < length - len + 1; i++) {
                if ( dict.contains(s.substring(i, i + len)) ) {
                    ref[i][len] = true;
                    continue;
                }
            
                for (int k = 1; k < len; k++) {
                    if (ref[i][k] && ref[k + i][len - k]) {
                        ref[i][len] = true;
                    }
                }
            }
        }
        if (!ref[0][length]) return ret;
        
        dfs(s, ref, 0, 0, length, new StringBuffer(), ret, dict);
        return ret;
    }
    
    private void dfs(String s, boolean[][] ref, int start, int depth, int length, StringBuffer sb, ArrayList<String> ret, Set<String> dict) {
        if (depth == length) {
            String t = sb.toString();
            ret.add(t.substring(0, t.length() - 1));
            return;
        }
        
        for (int i = 1; i <= length; i++) {
            if (ref[start][i]) {
                if ( dict.contains(s.substring(start, start + i)) ) {
                    int beforeAddLen = sb.length();
                    sb.append(s.substring(start, start + i)).append(" ");
                    dfs(s, ref, start + i, start + i, length, sb, ret, dict);
                    sb.delete(beforeAddLen, sb.length());
                }
            }
        }
        return;
    }
   
    
}

File: wordBreak2.cc
/* time limit exceed
class Solution {
public:
    vector<string> wordBreak(string s, unordered_set<string> &dict) {
        if (!s.size()) return vector<string>();
        return _break(s, dict, 0);
     }
    
    vector<string> _break(string &s, unordered_set<string> &dict, int start) {
        if (start >= s.size()) return vector<string>();
        vector<string> ret;
        for (int i = start; i < s.size(); i++) {
            vector<string> tmp;
            if ( dict.find(s.substr(start, i - start + 1)) != dict.end() ) {
                tmp = _break(s, dict, i + 1);
                for (int j = 0; j < tmp.size(); j++) {
                    tmp[j] = s.substr(start, i - start + 1) + " " + tmp[j];
                }
                ret.insert(ret.end(), tmp.begin(), tmp.end());
            }
        }
        return ret;
    }
     
};*/

class Solution {
public:
    vector<string> wordBreak(string s, unordered_set<string> &dict) {
        int n = s.length();
        if (!n) return vector<string> ();
        bool **ref = new bool*[n];
        for (int i = 0; i < n; i++) ref[i] = new bool[n + 1];
        vector<string> ret;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= n; j++)
                ref[i][j] = false;
        }
        for (int i = n - 1; i >= 0; i--) {
            for (int len = 1; len <= n - i; len++) {
                if (dict.count(s.substr(i, len)) > 0) 
                    ref[i][len] = true;
                
                for (int k = 1; k < len; k++) {
                    if (ref[i][k] && ref[i + k][len - k])
                        ref[i][len] = true;
                }
            }
        }
        if (!ref[0][n]) return vector<string>();
        string tmp = "";gi
        dfs(s, 0, tmp, dict, ref, ret);
        return ret;
    }
    
    void dfs(string s, int i, string tmp, unordered_set<string> &dict, bool **ref, vector<string> &ret) {
        int n = s.length();
        if (i == n) {
            tmp = tmp.substr(0, tmp.length() - 1);
            ret.push_back(tmp);
            return;
        }
        
        for (int len = 1; len <= n - i; len++) {
            if (ref[i][len]) {
                if (dict.count(s.substr(i, len))) {
                    int len_before = tmp.length();
                    tmp += s.substr(i, len);
                    tmp += " ";
                    dfs(s, i + len, tmp, dict, ref, ret);
                    tmp = tmp.substr(0, len_before);
                }
            }
        }
        return;
    }

    //-----------------------------------------------//
     vector<string> wordBreak(string s, unordered_set<string> &dict) {
        int n = s.length();
        if (!n) return vector<string>();
        vector<bool> tmp(n, false);
        vector<vector<bool>> dp(n, tmp);
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (dict.count(s.substr(i, j - i + 1)) > 0) {
                    dp[i][j] = true;
                    continue;
                }
                for (int k = i; k < j; k++) {
                    if (dp[i][k] && dp[k + 1][j]) {
                        dp[i][j] = true;
                        break;
                    }
                }
            }
        }
        if (!dp[0][n - 1]) return vector<string>();
        vector<string> ret;
        dfs(s, 0, ret, "", dp, dict);
        return ret;
    }
    
    void dfs(string &s, int start, vector<string> &ret, string tmp, vector<vector<bool>> &dp, unordered_set<string> &dict) {
        int n = s.length();
        if (start == n) {
            ret.push_back(tmp.substr(0, tmp.length() - 1));   
            return;
        }
        for (int len = 1; start + len <= n; len++) {
            if (dict.count(s.substr(start, len)) > 0) {
                if (dp[start][start + len - 1]) {
                    int len_before = tmp.length();
                    tmp += s.substr(start, len);
                    tmp += " ";
                    dfs(s, start + len, ret, tmp, dp, dict);
                    tmp = tmp.substr(0, len_before);
                }
            }
        }
        return;
    }
    
};
